mmf/models/transformers/base.py

```diff
@@ -0,0 +1,204 @@
+        specify the encoder as `self.encoder`, so for some transformer models this
+        method needs to be overriden in derived classes to initialize the encoder
+        properly.
+
+        Example ::
+
+            self.transformer = AutoModel.from_pretrained(
+                "bert-base-uncased",
+                config=self.transformer_config,
+            ).encoder
+        """
+        model = AutoModel.from_pretrained(
+            self.config.transformer_base,
+            config=self.transformer_config,
+            cache_dir=os.path.join(get_mmf_cache_dir(), "distributed_{}".format(-1)),
+        )
+        self.transformer = model.encoder
+
+    def build_heads(self):
+        """Build the different heads for the model. It can be either the pretraining
+        head or the classifier heads.
+
+        Example ::
+
+            # For pretraining
+            self.classifier = nn.Sequential(
+                BertPredictionHeadTransform(self.transformer_config),
+                nn.Linear(self.transformer_config.hidden_size, self.config.num_labels),
+            )
+
+            # For classification
+            self.classifier = nn.Sequential(
+                BertPredictionHeadTransform(self.transformer_config),
+                nn.Linear(self.transformer_config.hidden_size, self.config.num_labels),
+            )
+        """
+        pass
+
+    def build_losses(self):
+        """Initialize the losses for pretraining. For example MLM, MIM etc.
+
+        Example ::
+
+            self.mlm_loss = CrossEntropyLoss(ignore_index=-1)
+        """
+        pass
+
+    def _init_weights(self, module):
+        """Initialize the weights.
+        """
+        if isinstance(module, (nn.Linear, nn.Embedding)):
+            module.weight.data.normal_(mean=0.0, std=self.config.initializer_range)
+        elif isinstance(module, nn.module.LayerNorm):
+            module.bias.data.zero_()
+            module.weight.data.fill_(1.0)
+        if isinstance(module, nn.Linear) and module.bias is not None:
+            module.bias.data.zero_()
+
+    def tie_weights(self):
+        """Tie the weights between the input embeddings and the output embeddings
+        if required.
+        """
+        pass
+
+    def init_weights(self):
+        if self.config.random_initialize is False:
+            if self.config.transformer_base is None:
+                # No pretrained model, init weights
+                self.apply(self._init_weights)
+
+        # Tie weights if required
+        self.tie_weights()
+
+    def preprocess_sample(self, sample_list: Dict[str, Any]):
+        """Preprocess the sample_list. Generate image input ids, image position ids,
+        image mask. Create a attention mask for all the masks for different modalities.
+        """
+        # TODO(vedanuj): Make this method simpler
+        # Image feature(img_input_ids) and position(img_pos_ids) tokens
+        if "image_feature_0" in sample_list:
```

Comment:
Let's move this to a separate function for now so that its is easy to refactor later.

Reply:
This is removed from this base class for now